README.md
`seameless` is a TCP proxy that allow you to deploy new code then switch traffic
to it without downtime.

It does "round robin" between the list of current active backends.

Switching server is done with HTTP interface with the following API:

* `/set?backends=host:port,host:port`
    * set list of backends
* `/add?backend=host:port`
    * add a backend
* `/remove?backend=host:port`
    * remove a backend
* `/get`
    * return host:port,host:port

# Process
* Start `seamleass` with list of active backends

        seamless 8080 localhost:4444
* Direct all traffic to port 8080 on local machine.
* When you need to add/remove the backend, use the HTTP API on port 6777
  different port, say 4445)

        curl http://localhost:6777/add?backend=localhost:4445
        curl http://localhost:6777/remove?backend=localhost:4444

  Or

        curl http://localhost:6777/set?backends=localhost:4445
    
New traffic will be directed to new backend(s).

# Installing
You can download a statically linked executable at the [downloads][dl] section.

[dl]: https://bitbucket.org/tebeka/seamless/downloads

Or if you have a Go development environment, you can

    go get bitbucket.org/tebeka/seamless

# Contact
Miki Tebeka <miki.tebeka@gmail.com> or [here][prj].

[prj]: https://bitbucket.org/tebeka/seamless


# LICENSE
[MIT](https://bitbucket.org/tebeka/seamless/src/tip/LICENSE.txt)


ChangeLog
2014-06-19 version 0.2.6
    * Build with go 1.3

2013-12-02 version 0.2.5
    * Build with go 1.2.1

2013-12-02 version 0.2.4
    * Build with go 1.2

2013-06-14 version 0.2.3
    * Build with go 1.1.1

2012-09-06 version 0.2.2
    * Use lock in String

2012-09-06 version 0.2.1
    * backends.go (Thanks Fabrizio Milo)

2012-09-06 version 0.2.0
    * Support multiple backends [issue #3]

2012-09-03 version 0.1.2
    * Synchronize backend change [issue #2] (Thanks Jonathan Amsterdam)

2012-08-21 version 0.1.1
    * Handle error in http listener
    * Add -version flag

2012-08-10 version 0.1.0
    * Initial release


backends_test.go
package main

import (
	"testing"
)

// FIXME: Sync tests

func TestBackendsSet(t *testing.T) {
	bs := &Backends{}
	if len(bs.backends) != 0 {
		t.Fatalf("should be empty")
	}

	bs.current = 107
	bs.Set([]string{"a", "b"})
	if len(bs.backends) != 2 {
		t.Fatalf("len should 2 (was %d)", len(bs.backends))
	}

	if bs.current != 0 {
		t.Fatalf("current should be 0 (is %d)", bs.current)
	}
}

func TestBackendsNext(t *testing.T) {
	backend1, backend2 := "localhost:8888", "localhost:8887"
	bs := &Backends{backends: []string{backend1, backend2}}

	for i, expected := range []string{backend2, backend1, backend2} {
		next, _ := bs.Next()
		if next != expected {
			t.Fatalf("backend should be %s at %d (was %s)", expected, i, next)
		}
	}

	bs.Set([]string{})
	_, err := bs.Next()
	if err == nil {
		t.Fatalf("managed to get backend from empty list")
	}
}

func TestBackendsAdd(t *testing.T) {
	bs := &Backends{}
	bs.Add("a")
	if len(bs.backends) != 1 {
		t.Fatalf("not added")
	}
}

func TestBackendsRemove(t *testing.T) {
	count := -1

	bs := &Backends{}
	count = bs.Remove("a")
	if count != 0 {
		t.Fatalf("removed from empty list")
	}

	bs.Set([]string{"a", "b", "a"})
	count = bs.Remove("a")
	if count != 2 {
		t.Fatalf("should remove 2 (was %d)", count)
	}
	if len(bs.backends) != 1 {
		t.Fatalf("should have left one (have %d)", len(bs.backends))
	}

	if bs.backends[0] != "b" {
		t.Fatalf("wrong one left (%s)", bs.backends[0])
	}
}

func TestBackendsString(t *testing.T) {
	bs := &Backends{}
	if bs.String() != "" {
		t.Fatalf("bad empty str: %s", bs.String())
	}

	bs.Set([]string{"a", "b"})
	if bs.String() != "a,b" {
		t.Fatalf("bad str %s (expected a,b)", bs.String())
	}
}


seamless_test.go
// FIXME: Write a full test that starts a backend and seamless, and then
// switches
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"testing"
	"time"
)

var apiPort int = 6777
var numBackends int = 3
var proxyPort = 6888

type testHandler int

// startBackend spawns an HTTP server the listens on 6700+i port and replies with i to requests.
func (h testHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintf(w, "%d", h)
}

func startBackend(i int) {
	handler := testHandler(i)
	port := 6700 + i
	server := http.Server{Handler: handler, Addr: fmt.Sprintf(":%d", port)}
	go server.ListenAndServe()
}

func init() {
	for i := 0; i < numBackends; i++ {
		startBackend(i)
	}

	out := make(chan error)
	go seamless(fmt.Sprintf(":%d", proxyPort), apiPort, []string{}, out)

	time.Sleep(1 * time.Second)
}

func call(url string) (string, error) {
	// We really don't want keep alive or caching :)
	client := &http.Client{Transport: &http.Transport{DisableKeepAlives: true}}
	resp, err := client.Get(url)
	if err != nil {
		return "", fmt.Errorf("can't GET %s: %v\n", url, err)
	}
	defer resp.Body.Close()

	reply, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("error reading reply: %v\n", err)
	}

	return string(reply), nil
}

func callAPI(suffix string) (string, error) {
	url := fmt.Sprintf("http://localhost:%d/%s", apiPort, suffix)
	return call(url)
}

func TestHTTPGet(t *testing.T) {
	backends.Set([]string{"localhost:8080"})
	reply, err := callAPI("get")
	if err != nil {
		t.Fatalf("%s", err)
	}

	if reply != fmt.Sprintf("%s\n", backends.backends[0]) {
		t.Fatalf("bad reply: %s\n", string(reply))
	}
}

func TestHTTPAdd(t *testing.T) {
	backends.Set([]string{"localhost:8888"})
	backend := "localhost:8887"

	reply, err := callAPI(fmt.Sprintf("add?backend=%s", backend))
	if err != nil {
		t.Fatalf("%s", err)
	}

	if len(backends.backends) != 2 {
		t.Fatalf("bad number of backends (%d)\nreply: %s", len(backends.backends), reply)
	}

	if backends.backends[1] != backend {
		t.Fatalf("bad backend - %s", backends.backends[0])
	}

	if reply != fmt.Sprintf("%s,%s\n", backends.backends[0], backends.backends[1]) {
		t.Fatalf("bad reply - %s\n", reply)
	}
}

func TestHTTPRemove(t *testing.T) {
	backend1, backend2 := "localhost:8888", "localhost:8887"
	backends.Set([]string{backend1, backend2})
	reply, err := callAPI(fmt.Sprintf("remove?backend=%s", backend1))
	if err != nil {
		t.Fatalf("%s", err)
	}

	if len(backends.backends) != 1 {
		t.Fatalf("bad number of backends (%d)\nreply: %s", len(backends.backends), reply)
	}

	if backends.backends[0] != backend2 {
		t.Fatalf("bad backend left - %s", backends.backends[0])
	}
}

func Test_isValidBackend(t *testing.T) {
	names := map[bool]string{
		true:  "valid",
		false: "invalid",
	}

	cases := []struct {
		value string
		valid bool
	}{
		{"localhost:7", true},
		{"foo.com:8080", true},
		{"", false},
		{"foo.com", false},
		{"localhost", false},
		{"foo.com:", false},
	}

	for _, c := range cases {
		if isValidBackend(c.value) != c.valid {
			t.Fatalf("`%s` should be %s", c.value, names[c.valid])
		}
	}
}

func arreq(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}

	for i, v := range a {
		if b[i] != v {
			return false
		}
	}

	return true
}

func Test_parseBackends(t *testing.T) {
	cases := []struct {
		backends string
		expected []string
		ok       bool
	}{
		{"localhost:8080", []string{"localhost:8080"}, true},
		{"localhost:8080,localhost:8887", []string{"localhost:8080", "localhost:8887"}, true},
		{"", []string{}, false},
		{"foo", []string{}, false},
		{"localhost:8080,localhost", []string{}, false},
	}

	for _, c := range cases {
		value, err := parseBackends(c.backends)
		ok := err == nil

		if ok != c.ok {
			t.Fatalf("bad error for %v", c.backends)
		}

		if !arreq(value, c.expected) {
			t.Fatalf("go %v for %s (expected %v)", value, c.backends, c.expected)
		}

	}
}

func backendAddr(i int) string {
	return fmt.Sprintf("localhost:%d", 6700+i)
}

func callProxy() (string, error) {
	url := fmt.Sprintf("http://localhost:%d", proxyPort)
	return call(url)
}

func TestProxy(t *testing.T) {
	backends.Set([]string{backendAddr(0), backendAddr(1)})

	for i := 0; i < 7; i++ {
		reply, err := callProxy()
		if err != nil {
			t.Fatalf("can't call proxy - %v", err)
		}
		expected := fmt.Sprintf("%d", (i+1)%len(backends.backends))
		if reply != expected {
			t.Fatalf("bad backend for i=%d: got %s instead of %s", i, reply, expected)
		}
	}
}

func TestProxyRemove(t *testing.T) {
	backends.Set([]string{backendAddr(0), backendAddr(1)})
	suffix := fmt.Sprintf("remove?backend=%s", backendAddr(0))
	if _, err := callAPI(suffix); err != nil {
		t.Fatalf("can't remove %s - %s", backendAddr(0), err)
	}

	for i := 0; i < 7; i++ {
		reply, err := callProxy()
		if err != nil {
			t.Fatalf("can't call proxy - %v", err)
		}
		if reply != "1" {
			t.Fatalf("bad reply %s (expected 1)", reply)
		}
	}
}

func TestProxyAdd(t *testing.T) {
	backends.Set([]string{backendAddr(0), backendAddr(1)})

	suffix := fmt.Sprintf("add?backend=%s", backendAddr(2))
	if _, err := callAPI(suffix); err != nil {
		t.Fatalf("can't remove %s - %s", backendAddr(0), err)
	}

	for i := 0; i < 7; i++ {
		reply, err := callProxy()
		if err != nil {
			t.Fatalf("can't call proxy - %v", err)
		}
		expected := fmt.Sprintf("%d", (i+1)%len(backends.backends))
		if reply != expected {
			t.Fatalf("bad reply %s (expected %s)", reply, expected)
		}
	}
}


seesaw.go
/* Proxy TCP between two servers. Allow you to deploy new code to one server
   then switch traffic to it without downtime.

   Switching server is done with HTTP interface with the following API:
   /toggle - will toggle between servers
   /current - will return (in plain text) current server
   /servers - will return active, backup servers

Original code by Roger Peppe at http://bit.ly/Oc1YtF
*/
package main

import (
	"flag"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
)

var remotes = []string{"", ""}
var current int

func main() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "usage: seesaw LISTEN FIRST SECOND\n")
		flag.PrintDefaults()
	}
	port := flag.Int("httpPort", 6777, "http interface port")
	flag.Parse()
	if flag.NArg() != 3 {
		flag.Usage()
		os.Exit(1)
	}
	localAddr := flag.Arg(0)
	remotes[0] = flag.Arg(1)
	remotes[1] = flag.Arg(2)

	local, err := net.Listen("tcp", localAddr)
	if local == nil {
		die(fmt.Sprintf("cannot listen: %v", err))
	}

	go runHttpServer(*port)

	for {
		conn, err := local.Accept()
		if conn == nil {
			die(fmt.Sprintf("accept failed: %v", err))
		}
		go forward(conn, remotes[current])
	}
}

func forward(local net.Conn, remoteAddr string) {
	remote, err := net.Dial("tcp", remoteAddr)
	if remote == nil {
		fmt.Fprintf(os.Stderr, "remote dial failed: %v\n", err)
		return
	}
	go io.Copy(local, remote)
	go io.Copy(remote, local)
}

func die(msg string) {
	fmt.Fprintf(os.Stderr, "error: %s\n", msg)
	os.Exit(1)
}

func runHttpServer(port int) {
	http.HandleFunc("/toggle", toggleHandler)
	http.HandleFunc("/current", currentHandler)
	http.HandleFunc("/servers", serversHandler)
	http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
}

func toggleHandler(w http.ResponseWriter, req *http.Request) {
	current = 1 - current
	currentHandler(w, req)
}

func currentHandler(w http.ResponseWriter, req *http.Request) {
	fmt.Fprintf(w, "%s\n", remotes[current])
}

func serversHandler(w http.ResponseWriter, req *http.Request) {
	fmt.Fprintf(w, "%s\n", remotes[current])
	fmt.Fprintf(w, "%s\n", remotes[1-current])
}


backends.go
package main

import (
	"fmt"
	"strings"
	"sync"
)

// List of backends, support (thread safe) adding, removing and getting next in list (circular)
type Backends struct {
	backends []string
	current  int
	lock     sync.Mutex
}

// Set sets the current list of backends
func (bs *Backends) Set(backends []string) {
	bs.lock.Lock()
	defer bs.lock.Unlock()

	bs.backends = backends
	bs.current = 0
}

// Next returns the next back in circular fashion
func (bs *Backends) Next() (string, error) {
	bs.lock.Lock()
	defer bs.lock.Unlock()

	if len(bs.backends) == 0 {
		return "", fmt.Errorf("empty backends")
	}

	// We advance first to make sure we're in bounds
	bs.current = (bs.current + 1) % len(bs.backends)
	backend := bs.backends[bs.current]
	return backend, nil
}

// Add adds a new backend
func (bs *Backends) Add(backend string) {
	bs.lock.Lock()
	defer bs.lock.Unlock()

	bs.backends = append(bs.backends, backend)
}

// Remove removes all occurrences of backend from list of backends, returns the number of items removed
func (bs *Backends) Remove(backend string) int {
	bs.lock.Lock()
	defer bs.lock.Unlock()

	i, count := 0, 0
	for i < len(bs.backends) {
		if bs.backends[i] == backend {
			count++
			bs.backends = append(bs.backends[:i], bs.backends[i+1:]...)
		} else {
			i++
		}
	}

	return count
}

// String is string representation of backends
func (bs *Backends) String() string {
	bs.lock.Lock()
	defer bs.lock.Unlock()

	return strings.Join(bs.backends, ",")
}


seamless.go
/* A TCP proxy that allow you to deploy new code then switch traffic to it
   without downtime.

   It does "round robin" between the list of current active backends.

   Switching server is done with HTTP interface with the following API:
   /set?backends=host:port,host:port - will set list of backends
   /add?backend=host:port - will add a backend
   /remove?backend=host:port - will remove a backend
   /get - will return host:port,host:port

   Work flow:
	   Start first backend at port 4444
	   Run `./seamless 8080 localhost:4444`

	   Direct all traffic to port 8080 on local machine.

	   When you need to upgrade the backend, start a new one (with new code on
	   a different port, say 4445).
	   Then
			* `curl http://localhost:6777/add?backend=localhost:4445`
			* `curl http://localhost:6777/remove?backend=localhost:4444`
	   Or
		`curl http://localhost:6777/set?backends=localhost:4445`

	   New traffic will be directed to new server(s).

Original forward code by Roger Peppe (see http://bit.ly/Oc1YtF)
*/
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"regexp"
	"strings"
)

const (
	Version = "0.2.6"
)

// List of backends
var backends *Backends = &Backends{}

// backend regular expression (<host>:<port>)
var backendRe *regexp.Regexp = regexp.MustCompile("^[^:]+:[0-9]+$")

// isValidBackend returns true if backend is in "host:port" format
func isValidBackend(backend string) bool {
	return backendRe.MatchString(backend)
}

// parseBackends parses string in format "host:port,host:port" and return list of backends
func parseBackends(str string) ([]string, error) {
	backends := strings.Split(str, ",")
	if len(backends) == 0 {
		return nil, fmt.Errorf("no backends")
	}

	for i, v := range backends {
		backends[i] = strings.TrimSpace(v)
		if !isValidBackend(backends[i]) {
			return nil, fmt.Errorf("'%s' is not valid network address", backends[i])
		}
	}

	return backends, nil
}

// forward proxies traffic between local socket and remote backend
func forward(local net.Conn, remoteAddr string) {
	remote, err := net.Dial("tcp", remoteAddr)
	if err != nil {
		log.Printf("remote dial failed: %v\n", err)
		local.Close()
		return
	}
	go io.Copy(local, remote)
	go io.Copy(remote, local)
}

// die prints error message and aborts the program
func die(format string, args ...interface{}) {
	msg := fmt.Sprintf(format, args...)
	fmt.Fprintf(os.Stderr, "error: %s\n", msg)
	os.Exit(1)
}

// startHttpServer start the HTTP server interface in a given port
func startHttpServer(port int) error {
	http.HandleFunc("/set", setHandler)
	http.HandleFunc("/get", getHandler)
	http.HandleFunc("/add", addHandler)
	http.HandleFunc("/remove", removeHandler)

	return http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
}

// getHandler handles /current and return the current backend
func getHandler(w http.ResponseWriter, req *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintf(w, "%s\n", backends)
}

// setHandler handler /set and sets backends
func setHandler(w http.ResponseWriter, req *http.Request) {
	newBackends, err := parseBackends(req.FormValue("backends"))
	if err != nil {
		msg := fmt.Sprintf("error: %s", err)
		log.Println(msg)
		http.Error(w, msg, http.StatusBadRequest)
		return
	}

	backends.Set(newBackends)
	getHandler(w, req)
}

// addHandler handles /add to add a new backend
func addHandler(w http.ResponseWriter, req *http.Request) {
	backend := req.FormValue("backend")
	if len(backend) == 0 {
		msg := "error: missing 'backend' parameter"
		log.Println(msg)
		http.Error(w, msg, http.StatusBadRequest)
		return
	}

	backends.Add(backend)
	getHandler(w, req)
}

// removeHandler handles /remove and remove a backend
func removeHandler(w http.ResponseWriter, req *http.Request) {
	err := ""

	defer func() {
		if len(err) != 0 {
			log.Printf("error: %s\n", err)
			http.Error(w, err, http.StatusBadRequest)
			return
		} else {
			getHandler(w, req)
		}
	}()

	backend := req.FormValue("backend")
	if len(backend) == 0 {
		err = "missing 'backend' parameter"
		return
	}

	count := backends.Remove(backend)
	if count == 0 {
		err = fmt.Sprintf("backend '%s' not found", backend)
		return
	}
}

// seamless launches the HTTP API and then start proxying
func seamless(localAddr string, apiPort int, backendList []string, out chan error) {
	local, err := net.Listen("tcp", localAddr)
	if local == nil {
		out <- fmt.Errorf("cannot listen: %v", err)
		return
	}

	backends.Set(backendList)

	go func() {
		if err := startHttpServer(apiPort); err != nil {
			out <- fmt.Errorf("cannot listen on %d: %v", apiPort, err)
		}
	}()

	for {
		conn, err := local.Accept()
		if conn == nil {
			die("accept failed: %v", err)
		}
		backend, err := backends.Next()
		if err != nil {
			log.Printf("error: can't get next backend %v\n", err)
			conn.Close()
		}
		go forward(conn, backend)
	}
}

func main() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "usage: seamless LISTEN_PORT BACKENDS\n")
		fmt.Fprintf(os.Stderr, "command line switches:\n")
		flag.PrintDefaults()
	}
	port := flag.Int("httpPort", 6777, "http interface port")
	version := flag.Bool("version", false, "show version and exit")
	flag.Parse()

	if *version {
		fmt.Printf("seamless %s\n", Version)
		os.Exit(0)
	}

	if flag.NArg() != 2 {
		flag.Usage()
		os.Exit(1)
	}
	localAddr := fmt.Sprintf(":%s", flag.Arg(0))

	var err error
	backendList, err := parseBackends(flag.Arg(1))
	if err != nil {
		die(fmt.Sprintf("%s", err))
	}

	out := make(chan error)
	go seamless(localAddr, *port, backendList, out)

	err = <-out
	if err != nil {
		die("%s", err)
	}
}
